Fix batch split bug

heppu opened the pull request:
Parsing breaks on `GOTO` word so I added ahead lookup with length validation to fix this. Without this fix running DB migration scripts isn't possible. PR waiting for this [here](https://github.com/golang-migrate/migrate/pull/1270).

heppu committed:
Add test for batch.Split bug

heppu committed:
Fix GOTO batch parsing bug

heppu commented:
@microsoft-github-policy-service agree

heppu commented:
@shueybubbles could you take a look at this?

shueybubbles commented on batch/batch.go:
```diff
@@ -46,6 +46,12 @@ func hasPrefixFold(s, sep string) bool {
 	}
+
+	// Forward lookup to see if the word continues.
+	if len(s) > len(sep) && unicode.IsLetter(rune(s[len(sep)])) {
+		return false
```
@heppu add some double byte char test cases to cover this.

dlevy-msft-sql committed:
fix: use DecodeRuneInString for word-boundary check

Addresses review feedback on the GO/GOTO word-boundary fix:

- Use utf8.DecodeRuneInString so the follower-char letter check sees the full rune, not the leading byte of a multi-byte UTF-8 sequence. Casting rune(byte) misclassifies multi-byte runes (for example Hebrew aleph U+05D0 has leading byte 0xD7 which is the MULTIPLICATION SIGN, not a letter).

- Extend TestHasPrefixFold to cover word-boundary cases (GOTO, gotoflag, GO1, GO_FOO), Latin-1 follower, and the Hebrew-aleph case that distinguishes the two implementations.

Co-authored-by: Henri Koski <henri.t.koski@gmail.com>

dlevy-msft-sql committed:
merge: resolve upstream/main into fix/batch-split-bug

Keep both new TestBatchSplit cases: the GOTO/Bookmark case from this branch and the create-table 'gone_ts' case from #248 on main. Both exercise the word-boundary protection from different angles.
